### PR TITLE
Tangent-frame shear loss decomposition to reduce wsy/wsz error

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Wall-shear (wsy, wsz) predictions are structurally worse than surface-pressure — the current best has wsy=8.73% vs target 3.65% (5.08pp gap) and wsz=10.58% vs target 3.63% (6.95pp gap). The model predicts shear in Cartesian (x,y,z) space, but shear stress is fundamentally a **tangential field** — it should lie in the tangent plane of the surface. By decomposing the shear-loss into tangential and normal components *in a local tangent frame*, we can impose a geometric constraint that forces the model's shear predictions to stay on-manifold. This decomposition may reduce the structural bias that inflates wsy/wsz errors.

**Mechanism**: For each surface point, build a local tangent frame {t1, t2, n} from the surface normals `sdf_normals` already available in the batch. Project the predicted and target wall-shear vectors onto this frame. Compute separate MSE contributions for (a) the tangential components (high-weight, geometrically meaningful) and (b) the normal component (low-weight or near-zero, should be ~0). The ratio of tangential to normal energy serves as an auxiliary regularization signal.

Reference: standard CFD post-processing decomposes wall-shear into tangential components routinely. This imposes a physics-informed inductive bias on the output space.

## Instructions

**Train on the 5L/256d + FourierEmbed + T_max=50 baseline recipe** and add a tangent-frame loss decomposition for wall-shear:

### Step 1 — Locate surface normals in the batch

In `train.py`, inspect the batch dict to find the key that provides surface point normals. It is likely `sdf_normals` or similar. These are unit normals at each surface point. Print one batch's keys at ep0 to confirm. If normals are unavailable, use finite-difference estimation from the SDF.

### Step 2 — Compute per-point tangent frames

For each surface point with normal `n` (shape [N, 3]):
```python
def gram_schmidt_tangent_frame(n):
    # n: [N, 3], unit normals
    # Pick a random orthogonal seed (avoid collinearity with n)
    seed = torch.zeros_like(n)
    seed[:, 0] = 1.0
    collinear = (n[:, 0].abs() > 0.9)
    seed[collinear, 0] = 0.0
    seed[collinear, 1] = 1.0
    # Gram-Schmidt
    t1 = seed - (seed * n).sum(-1, keepdim=True) * n
    t1 = F.normalize(t1, dim=-1)
    t2 = torch.cross(n, t1, dim=-1)
    return t1, t2  # both [N, 3]
```

### Step 3 — Decompose shear into tangential/normal components

For predicted `shear_pred` [N, 3] and target `shear_tgt` [N, 3]:
```python
t1, t2 = gram_schmidt_tangent_frame(normals)
pred_t1 = (shear_pred * t1).sum(-1)   # [N]
pred_t2 = (shear_pred * t2).sum(-1)   # [N]
pred_n  = (shear_pred * normals).sum(-1)  # [N] — should be ~0

tgt_t1  = (shear_tgt * t1).sum(-1)
tgt_t2  = (shear_tgt * t2).sum(-1)
# tgt_n ≈ 0 by definition (shear is tangential)

loss_tangential = F.mse_loss(pred_t1, tgt_t1) + F.mse_loss(pred_t2, tgt_t2)
loss_normal_pen = pred_n.pow(2).mean()  # penalty for normal leakage

tangent_loss = loss_tangential + 0.1 * loss_normal_pen
```

Add `tangent_loss` to the overall training loss with a tunable weight `--tangent-frame-weight`:
- Test **two values**: `w=0.1` and `w=0.5` (run them sequentially in the same PR).
- Use `--wandb-group bengio-wave17-emma` so both runs are grouped.

### Step 4 — Training command

**Trial A (w=0.1)**:
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 50 \
  --no-compile-model \
  --tangent-frame-weight 0.1 \
  --wandb-group bengio-wave17-emma
```

**Trial B (w=0.5)** — run after Trial A completes (or in parallel if a second GPU slot is available):
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 50 \
  --no-compile-model \
  --tangent-frame-weight 0.5 \
  --wandb-group bengio-wave17-emma
```

### Step 5 — Expected code changes

Modify `target/train.py`:
1. Add `--tangent-frame-weight` argument (type=float, default=0.0).
2. Extract surface normals from the batch (check available keys).
3. Implement `gram_schmidt_tangent_frame()` as a local helper.
4. Inside the loss computation, if `tangent_frame_weight > 0`, extract the wall-shear component of predictions/targets, decompose, and add weighted tangent loss to the main loss.
5. Log `train/tangent_loss` and `train/normal_leakage` to W&B for diagnostics.

If `sdf_normals` or equivalent is not available in the batch, fall back to: use the **z-axis** as approximate normal (flat-plate approximation) for the tangential decomposition. This is a coarser proxy but will still bias predictions toward the tangential manifold.

### Kill/pass thresholds

| Gate | Epoch | Kill if abupt ≥ |
|------|-------|-----------------|
| ep5  | 5 | 13.0% |
| ep10 | 10 | 11.0% |
| ep15 | 15 | 9.5% |
| ep30 | 30 | 8.0% |

Primary success criterion: `val_primary/abupt_axis_mean_rel_l2_pct < 6.9549%` (current best). **Key metric to watch**: wsy and wsz — did the tangent-frame decomposition reduce the 5-7pp binding gaps?

## Baseline

Current best (alphonse PR #174, W&B run `vu4jsiic`, ep~45.3, step 807,025):

| Metric | Current Best (val) | AB-UPT Target |
|--------|-------------------|--------------|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **6.9549** | 4.51 |
| `val_primary/surface_pressure_rel_l2_pct` | 4.5644 | 3.82 |
| `val_primary/volume_pressure_rel_l2_pct` | 3.9361 ✓ | 6.08 |
| `val_primary/wall_shear_y_rel_l2_pct` | 8.7345 | 3.65 |
| `val_primary/wall_shear_z_rel_l2_pct` | 10.5766 | 3.63 |

Reproduce command for baseline:
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 50 \
  --no-compile-model \
  --wandb-group bengio-wave9
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: @emma — This Wave 17 PR was assigned on 2026-05-02 and has had zero activity. Your pod is 1/1 READY. Please ACK this PR within the next hour and confirm you have launched your run. Post your run ID (wandb run URL or run-id hash) as a comment here once started.

If you have a blocker or question, post it now and I will respond immediately. Unacknowledged assignments will be reassigned.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #475 — nudge posted at 03:39Z (over 2 hours ago), no ACK received. Pod confirmed as zombie in issue #466. Will reassign to @emma once human team performs SIGKILL and pod recovers.
